inc/c_util/tqueue_ll.h

```diff
@@ -378,13 +385,52 @@ TQUEUE_POP_RESULT TQUEUE_LL_POP(C)(TQUEUE_LL(T) tqueue, T* item, void* copy_item
+    {                                                                                                                                                               \
+        LogError("Invalid arguments: TQUEUE_LL(" MU_TOSTRING(T) ") tqueue=%p.", tqueue);                                                                            \
+        result = 0;                                                                                                                                                 \
+    }                                                                                                                                                               \
+    else                                                                                                                                                            \
+    {                                                                                                                                                               \
+        int64_t current_tail = 0;                                                                                                                                   \
+        int64_t current_head = 0;                                                                                                                                   \
+        do                                                                                                                                                          \
+        {                                                                                                                                                           \
+            /* Codes_SRS_TQUEUE_22_003: [ TQUEUE_GET_VOLATILE_COUNT(T) shall obtain the current tail queue by calling interlocked_add_64. ]*/                       \
+            current_tail = interlocked_add_64((volatile_atomic int64_t*)&tqueue->tail, 0);                                                                          \
+            /* Codes_SRS_TQUEUE_22_002: [ TQUEUE_GET_VOLATILE_COUNT(T) shall obtain the current head queue by calling interlocked_add_64. ]*/                       \
+            current_head = interlocked_add_64((volatile_atomic int64_t*)&tqueue->head, 0);                                                                          \
+            /* Codes_SRS_TQUEUE_22_006: [ TQUEUE_GET_VOLATILE_COUNT(T) shall obtain the current tail queue again by calling interlocked_add_64 and compare with the previosuly obtained tail value.  The tail value is valid only if it has not changed. ]*/   \
+        } while (current_tail != interlocked_add_64((volatile_atomic int64_t*) & tqueue->tail, 0));                                                                 \
+                                                                                                                                                                    \
```

Comment:
I have to question why do we really want a loop?
Frankly the first values we compute are as good as the 2nd one or the next one.

The loop achieves "let's obtain the head while we know the tail has not changed".
But it won't achieve anyway "let's obtain the tail while the head does not change".

And any of them can change the microsecond after we exit the loop anyway...

So, simplicity might be better here, just saying.

Reply:
The while loop is to ensure we returns the correct count at a point of time.  Consider the following scenario without the loop (only get tail and head once):
assume tail at 1 and head at 2 => 1 item in queue
1) get tail => 1
2) some other threads execute and do push+pop+push+pop+push+pop.... which advanced tail to 100 and head to 101
3) get head => 101
Above would have get_count return 100.  But actually, the max items in queue at any time during the above period was 2.

Reply:
Makes sense, it is better, basically tail is some sort of invariant.